group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -8540,5 +8568,1182 @@ private static class RebalanceResult {
+        CoordinatorResult<Void, Record> result = context.sendGenericGroupSync(syncRequest, syncFuture);
+
+        assertEquals(
+            Collections.singletonList(RecordHelpers.newGroupMetadataRecord(group, MetadataVersion.latest())),
+            result.records()
+        );
+        // Simulate a successful write to the log.
+        result.appendFuture().complete(null);
+
+        assertTrue(syncFuture.isDone());
+        assertEquals(Errors.NONE.code(), syncFuture.get().errorCode());
+        assertTrue(group.isInState(STABLE));
+
+        // Add a new dynamic pending member.
+        CompletableFuture<JoinGroupResponseData> joinFuture = new CompletableFuture<>();
+        result = context.sendGenericGroupJoin(
+            joinRequest.setMemberId(UNKNOWN_MEMBER_ID)
+                .setGroupInstanceId(null)
+                .setSessionTimeoutMs(5000),
+            joinFuture,
+            true,
+            true
+        );
+
+        assertTrue(result.records().isEmpty());
+        assertTrue(joinFuture.isDone());
+        assertEquals(Errors.MEMBER_ID_REQUIRED.code(), joinFuture.get().errorCode());
+        assertEquals(1, group.numPendingJoinMembers());
+        assertTrue(group.isInState(STABLE));
+
+        // Heartbeat from the leader to maintain session while timing out pending member.
+        HeartbeatRequestData heartbeatRequest = new HeartbeatRequestData()
+            .setGroupId("group-id")
+            .setMemberId(firstMemberId)
+            .setGenerationId(firstGenerationId);
+
+        for (int i = 0; i < 2; i++) {
+            assertNoOrEmptyResult(context.sleep(2500));
+            HeartbeatResponseData heartbeatResponse = context.sendGenericGroupHeartbeat(heartbeatRequest);
+            assertEquals(Errors.NONE.code(), heartbeatResponse.errorCode());
+        }
+
+        // At this point the second member should have been removed from pending list (session timeout),
+        // and the group should be in Stable state with only the first member in it.
+        assertEquals(1, group.size());
+        assertTrue(group.hasMemberId(firstMemberId));
+        assertEquals(1, group.generationId());
+        assertTrue(group.isInState(STABLE));
+    }
+
+    @Test
+    public void testRebalanceTimesOutWhenSyncRequestIsNotReceived() throws Exception {
+        // This test case ensure that the pending sync expiration does kick out all members
+        // if they don't send sync requests before the rebalance timeout. The
+        // group is in the Empty state in this case.
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .build();
+        GenericGroup group = context.createGenericGroup("group-id");
+
+        int rebalanceTimeoutMs = 5000;
+        int sessionTimeoutMs = 5000;
+        List<JoinGroupResponseData> joinResponses = joinWithNMembers(context, group, 3, rebalanceTimeoutMs, sessionTimeoutMs);
```

Comment:
Should we verify that the responses are OK?

Reply:
The join responses are verified inside `joinWithNMembers`